framework/include/image_resource.hpp

```diff
@@ -0,0 +1,277 @@
+#pragma once
+#include <gvk.hpp>
+
+namespace gvk
+{
+	class image_resource_t;
+	class image_resource_impl_t;
+
+	typedef std::unique_ptr<image_resource_t> image_resource;
+
+	// interface of image_resource type, used for abstraction and implementor in bridge pattern
+	class image_resource_base_t
+	{
+	public:
+		// type that represents the size of 1D, 2D, and 3D images
+		typedef vk::Extent3D extent_type;
+
+		// load image resource into memory
+		// perform this as an extra step to facilitate caching of image resources
+		virtual void load() = 0;
+
+		virtual vk::Format get_format() const = 0;
+
+		virtual vk::ImageType target() const = 0;
+
+		// size of image resource, in pixels
+		virtual extent_type extent(uint32_t level = 0) const = 0;
+
+		// Note: the return type cannot be const void* because this would result in a compile-time error with the deserializer;
+		// The function cannot be const either or gli would call a function overload that returns a const void*
+		virtual void* get_data(uint32_t layer, uint32_t face, uint32_t level) = 0;
+		// size of whole image resource, in bytes
+		virtual size_t size() const = 0;
+		// size of one mipmap leve of image resource, in bytes
+		virtual size_t size(uint32_t level) const = 0;
+
+		virtual bool empty() const = 0;
+
+		// Vulkan uses uint32_t type for levels and layers (faces)
+		virtual uint32_t levels() const = 0;
+
+		// array layers, for texture arrays
+		virtual uint32_t layers() const = 0;
+
+		// faces in image, must be 6 for cubemaps, 1 for anything else
+		virtual uint32_t faces() const = 0;
+
+		// returns true if image format and image library support y-flipping when loading
+		// TODO: make protected or remove?
+		virtual bool can_flip() const = 0;
+
+		// returns if image resource is a hdr format
+		virtual bool is_hdr() const = 0;
+
+		std::string path() const
+		{
+			return mPaths[0];
+		}
+
+		std::vector<std::string> paths() const
+		{
+			return mPaths;
+		}
+
+	protected:
+		
+		image_resource_base_t(const std::string& aPath, bool aHDR = false, bool asRGB = false, bool aFlip = false, int aPreferredNumberOfTextureComponents = 4)
```

Comment:
The meaning is the same, but shorter variable names may increase readability of the code. The exact meaning of the parameters will be explained in the documentation anyhow. I have changed the names for now to align with the rest of the code, however the current names might be misleading:
for textures loaded with GLI, both HDR and sRGB parameters are ignored; the image format only depends on the texture file. For stbi, HDR = true results in a (crude) conversion of 8-bit textures to float16 type, and vice versa for HDR = false. The implementation doesn't actually check the if the texture file is HDR first. Similarly, with stbi the texture data is treated as sRGB solely based on the value of the sRGB parameter. This behaviour is unchanged between the main branch and the cubemap branch.
I think it would be favorable to change this behaviour and the meaning of the parameters, but only after the cubemap implementation is merged.

Reply:
Okay. That's fine. I agree that the selection of the configuration parameters and their use everywhere might not be optimal. I would welcome a future restructuring regarding them. 

Having the same parameter _naming_ across the whole framework has the advantage of consistency: If a framework user knows the meaning of a parameter once, he/she can assume the same meaning if a parameter of exactly the same name is encountered at another place.

Reply:
As already mentioned in another comment, the Config or Meta pattern could probably be used to have only one function create_image_*  for regular images and cubemaps, and make all parameters optional, thus reducing the number of user-visible functions.